observer/probers/tls/tls.go

```diff
@@ -0,0 +1,193 @@
+package probers
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/crypto/ocsp"
+)
+
+type reason int
+
+type badOutcomeError struct {
+	reason reason
+}
+
+const (
+	none reason = iota
+	internalError
+	issuerVerifyFailed
+	ocspUnknown
+	ocspError
+)
+
+var reasonToString = map[reason]string{
+	none:               "nil",
+	internalError:      "internalError",
+	issuerVerifyFailed: "issuerVerifyFailed",
+	ocspUnknown:        "ocspUnknown",
+	ocspError:          "ocspError",
+}
+
+func (e badOutcomeError) Error() string {
+	return reasonToString[e.reason]
+}
+
+func getReasons() []string {
+	var allReasons []string
+	for _, v := range reasonToString {
+		allReasons = append(allReasons, v)
+	}
+	return allReasons
+}
+
+// TLSProbe is the exported `Prober` object for monitors configured to perform
+// TLS protocols.
+type TLSProbe struct {
+	url        string
+	rootOrg    string
+	rootCN     string
+	response   string
+	certExpiry *prometheus.GaugeVec
+	reason     *prometheus.GaugeVec
+}
+
+// Name returns a string that uniquely identifies the monitor.
+func (p TLSProbe) Name() string {
+	return fmt.Sprintf("%s-expecting-%s-%s-%s", p.url, p.response, p.rootOrg, p.rootCN)
+}
+
+// Kind returns a name that uniquely identifies the `Kind` of `Prober`.
+func (p TLSProbe) Kind() string {
+	return "TLS"
+}
+
+// Get OCSP status (good, revoked or unknown) of certificate
+func getOCSPStatus(cert, issuer *x509.Certificate) (int, error) {
+	req, err := ocsp.CreateRequest(cert, issuer, nil)
+	if err != nil {
+		return ocsp.Unknown, err
+	}
+
+	url := fmt.Sprintf("%s/%s", cert.OCSPServer[0], base64.StdEncoding.EncodeToString(req))
+	res, err := http.Get(url)
+	if err != nil {
+		return ocsp.Unknown, err
+	}
+
+	output, err := io.ReadAll(res.Body)
+	if err != nil {
+		return ocsp.Unknown, err
+	}
+
+	ocspRes, err := ocsp.ParseResponseForCert(output, cert, issuer)
+	if err != nil {
+		return ocsp.Unknown, err
+	}
+
+	switch ocspRes.Status {
+	case ocsp.Revoked:
+		return ocsp.Revoked, nil
+	case ocsp.Good:
+		return ocsp.Good, nil
+	default:
+		return ocsp.Unknown, nil
+	}
+}
+
+// Make a TLS connection with InsecureSkipVerify set to true to get the
+// expiration time and root information of insecure leaf certificates.
+func (p TLSProbe) getInsecureCertInfo() (string, string, time.Duration, error) {
+	conf := &tls.Config{InsecureSkipVerify: true}
+	conn, err := tls.Dial("tcp", p.url+":443", conf)
+	if err != nil {
+		return "", "", 0, err
+	}
+	peers := conn.ConnectionState().PeerCertificates
+	expiryDate := peers[0].NotAfter
+
+	for i := 0; i < len(peers)-1; i++ {
+		certIssuer, issuerSubject := peers[i].Issuer, peers[i+1].Subject
+		if (certIssuer.CommonName != issuerSubject.CommonName) || (certIssuer.Organization[0] != issuerSubject.Organization[0]) {
+			err = fmt.Errorf(reasonToString[issuerVerifyFailed])
```

Comment:
Two comments here:
1) In general, assigning a value inside a loop like this, and then simply going on with the next iteration of the loop, isn't the best behavior. If we've run into an error already, we should probably break out and return it, rather than overwriting it with a pootentially-slightly-different error next loop. So I'd immediately return the error after this.
2) This can just be `err = badOutcomeError{issuerVerifyFailed}`